### DISCUSSION TITLE
Representing WCHAR and CHAR types on structs

### DISCUSSION THREAD

**Discussion by sotteson1** [microsoft/win32metadata#394] [Ideas]
We're using PWSTR and PSTR structs to wrap pointers to strings. They are defined as:

```C#
[NativeTypedef]
public struct PWSTR
{
    public unsafe char* Value;
}

[NativeTypedef]
public struct PSTR
{
    public unsafe byte* Value;
}
```
This lets a projection know that a field/parameter of this type is a string, and not just a pointer to a single char or byte. Now I need to do something about fixed-length strings that appear on structs. I can't use PWSTR and PSTR because these aren't pointers to strings--they are a fixed number of bytes which contain strings:

```C#
public struct WIN32_FIND_DATAW
{
...
    public char[260] cFileName;

    public char[14] cAlternateFileName;
}

public struct WIN32_FIND_DATAA
{
...
    public byte[260] cFileName;

    public byte[14] cAlternateFileName;
}
```

The W version isn't bad because "char" implies that it's a string, but ECMA-335 doesn't have an equivalent for single byte characters, so it's not obvious that WIN32_FIND_DATAA.cFileName is meant to contain single byte characters. The original definition for WIN32_FIND_DATAA uses CHAR which lets the developer know what the field is for:

```C
public struct WIN32_FIND_DATAA
{
...
    public CHAR cFileName[260];

    public CHAR cAlternateFileName[14];
}
```

I could make a new struct called CHAR and use it in places like WIN32_FIND_DATAA:

```C#
[NativeTypedef]
public struct CHAR
{
    public byte Value;
}

public struct WIN32_FIND_DATAA
{
...
    public CHAR[260] cFileName;

    public CHAR[14] cAlternateFileName;
}
```

To the projection authors @kennykerr, @timsneath and @AArnott, what do you think?

**Comment by AArnott** [comment]
`CHAR` is fine, if ECMA lets you define fixed length inline arrays with a custom struct. C# doesn't, so CsWin32 will have to do magic to workaround that. But as you effectively convey the semantic this way, it seems adequate.

The only alternative I can think of is to use `byte[260]` and add an attribute identifying it as an 8-bit char string.

For 16-bit characters, UTF-16 is well understood when using `char` in ECMA-355.
How can we tell what the encoding is meant to be when `CHAR` is used? Is it ANSI or is it UTF-8? Should we make it more explicit in the typedef struct name and define potentially multiple of them? As the projections may want to translate a native string for a given runtime/platform to the `CHAR` used here, we need to know which encoding to use. If all of them are just one encoding, then let's just document it and go with what you propose. But if it may vary by API, we should consider multiple typedefs IMO.

**Reply by sotteson1** [reply]
There aren't different Win32 structs/APIs for UTF-8. How the OS interprets the -A APIs is determined by the process code page:

https://docs.microsoft.com/en-us/windows/uwp/design/globalizing/use-utf8-code-page
"Until recently, Windows has emphasized "Unicode" -W variants over -A APIs. However, recent releases have used the ANSI code page and -A APIs as a means to introduce UTF-8 support to apps. If the ANSI code page is configured for UTF-8, -A APIs operate in UTF-8. This model has the benefit of supporting existing code built with -A APIs without any code changes."

**Comment by timsneath** [comment]
I'm a bit unclear right now as to the right path forward for the Dart projection, so this thread is interesting to me. So far I've chosen to only wrap the Unicode APIs, so I don't yet have an opinion on the ANSI/UTF-8 APIs. 

Ideally I'd prefer not to expose both APIs -- in the same way that C allows you to set `#define UNICODE` and then call something like `ChooseFont` rather than `ChooseFontW`, I'd like to hide some of the complexity here. 

One obstacle I have with `struct`s is that there's no programmatic way to identify the 'unencumbered' name of the struct. I'd like to be able to interrogate an attribute to discover that `WIN32_FIND_DATAA` is the ANSI version of the `WIN32_FIND_DATA` struct. Of course, I can't just read the last letter, because the `A` might just be the last letter of the name itself, rather than a suffix. Today I have had to add a JSON file that provides a mapping between the ILSpy name and its "canonical" representation, but ideally I'd retrieve this from the metadata.

Hope this helps!

**Reply by sotteson1** [reply]
Thanks, that's helpful. @AArnott, aren't you doing something similar in CSWin32, where someone can just say they want CreateFile and under the covers they're using CreateFileW? How do you do that? It seems like data that should be in the metadata but I haven't thought how I would implement it yet. Maybe you already have a good idea for it.

**Reply by AArnott** [reply]
In CsWin32 I follow a heuristic of a trailing `A` or `W` character. I default to omitting generation of the `A` ones and rename the `W` ones to remove the suffix (to match the experience in C). Users *can* enable generating both, at which point I keep the suffix for A and W functions, also to match C.

Relevant code snippets from CsWin32:

https://github.com/microsoft/CsWin32/blob/553bae39c3eca9a2ecd2259cd66f89495affdc23/src/Microsoft.Windows.CsWin32/Generator.cs#L729-L740

https://github.com/microsoft/CsWin32/blob/553bae39c3eca9a2ecd2259cd66f89495affdc23/src/Microsoft.Windows.CsWin32/Generator.cs#L1858-L1882

The `[DllImport]` attributes in the metadata could set the `CharSet` property as the way to indicate this. But not all W methods have A counterparts and vice versa, so the value of CharSet alone should probably not be used to filter out methods since we probably want at least one charset of every method in the generated code. So if the goal is for the metadata to help projections omit A methods where W methods also exist, that will either continue to require special consideration in the projection or in the metadata.

**Reply by timsneath** [reply]
Curious -- what do you do with something like `Windows.Win32.Mib.MIB_BOUNDARYROW`? How do you know whether this struct is called `MIB_BOUNDARY_ROW` or is a Unicode struct called `MIB_BOUNDARY_RO`?

**Reply by AArnott** [reply]
I don't check *structs* for the suffix. I only check methods, so such an ambiguity is less likely to exist (and doesn't, AFAIK).
I only generate structs that support methods that are actually generated (or that are explicitly requested in NativeMethods.txt), so you implicitly don't get `A` structs that aren't used by generated methods.
I also never rename structs as I may do for methods.

**Reply by timsneath** [reply]
Got it. So your `ChooseFont` method takes a `CHOOSEFONTW` as a parameter?

**Reply by AArnott** [reply]
Yes.

**Reply by timsneath** [reply]
Ugh. I was hoping you'd come up with some magic that I hadn't yet divined :) OK, @sotteson1, I think you have an opportunity for a customer delighter feature here :)
